wdl-grammar/src/tree/dive.rs

```diff
@@ -0,0 +1,265 @@
+    /// [`SyntaxElement`](rowan::SyntaxElement)s for which the `match_predicate`
+    /// evaluates to `true`. No children nodes will be searched for
+    /// [`SyntaxNode`](rowan::SyntaxNode)s that are visited and for which the
+    /// `ignore_predicate` evaluates to `true`. In this way, you can ignore
+    /// entire sections of the tree.
+    ///
+    /// No guarantees are made about the order in which elements will be
+    /// traversed or collected.
+    fn dive_with_ignore<M, I>(
+        &self,
+        match_predicate: M,
+        ignore_predicate: I,
+    ) -> impl Iterator<Item = rowan::SyntaxElement<L>>
+    where
+        M: Fn(&rowan::SyntaxElement<L>) -> bool,
+        I: Fn(&rowan::SyntaxNode<L>) -> bool;
+}
+
+impl<D, L> Divable<L> for &D
+where
+    D: Divable<L>,
+    L: Language,
+{
+    fn dive_with_ignore<M, I>(
+        &self,
+        match_predicate: M,
+        ignore_predicate: I,
+    ) -> impl Iterator<Item = rowan::SyntaxElement<L>>
+    where
+        M: Fn(&rowan::SyntaxElement<L>) -> bool,
+        I: Fn(&rowan::SyntaxNode<L>) -> bool,
+    {
+        D::dive_with_ignore(self, match_predicate, ignore_predicate)
+    }
+}
+
+impl Divable<WorkflowDescriptionLanguage> for SyntaxNode {
+    fn dive_with_ignore<M, I>(
+        &self,
+        match_predicate: M,
+        ignore_predicate: I,
+    ) -> impl Iterator<Item = SyntaxElement>
+    where
+        M: Fn(&SyntaxElement) -> bool,
+        I: Fn(&SyntaxNode) -> bool,
+    {
+        DiveIterator::new(
+            // NOTE: this is an inexpensive clone of a red node.
+            SyntaxElement::Node(self.clone()),
+            match_predicate,
+            ignore_predicate,
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::OnceLock;
+
+    use rowan::GreenNode;
+
+    use crate::dive::Divable;
+    use crate::SyntaxKind;
+    use crate::SyntaxNode;
+    use crate::SyntaxTree;
+
+    fn get_syntax_node() -> SyntaxNode {
+        static GREEN_NODE: OnceLock<GreenNode> = OnceLock::new();
+
+        let green_node = GREEN_NODE
+            .get_or_init(|| {
+                let (tree, diagnostics) = SyntaxTree::parse(
+                    r#"version 1.2
+    
+task hello {
+    String a_private_declaration = false
+}
+
+workflow world {
+    String another_private_declaration = true
+}"#,
+                );
+
+                assert!(diagnostics.is_empty());
+                tree.green()
+            })
+            .clone();
+
+        SyntaxNode::new_root(green_node)
+    }
+
+    #[test]
+    fn it_dives_correctly() {
+        let tree = get_syntax_node();
+
+        let mut ignored_idents = tree.dive(|element| element.kind() == SyntaxKind::Ident);
+
+        // NOTE: unlike the visitor, which carries out a preorder traversal, the
+        // current implementation of dive in this create uses a level-order
+        // traversal. Thus, the names of the workflow and task will be seen
+        // before their inner members.
```

Comment:
Do you mean the names of **all** workflow and tasks will be seen before their members?

Reply:
In this case, yes. The idea is that a breadth-first search is carried out in this iterator, so the `workflow` and `task` nodes will be visited before their children.